src/storage/core.rs

```diff
@@ -0,0 +1,471 @@
+    S: SpawnExt + Send + 'static + Unpin,
+{
+    type Item = FutureObj<'static, Result<()>>;
+
+    fn poll_next(mut self: Pin<&mut Self>, waker: &Waker) -> Poll<Option<Self::Item>> {
+        let now = Instant::now();
+        if self.next_update < now {
+            trace!("Observer update");
+            self.as_mut().next_update = now + self.update_interval;
+            let storage_cloned = self.storage.clone();
+            let fut = update_active_blob(storage_cloned);
+            let fut_boxed = Box::new(fut);
+            let obj = FutureObj::new(fut_boxed);
+            Poll::Ready(Some(obj))
+        } else if self.storage.shared.observer_state.load(Ordering::Relaxed) {
+            waker.wake();
+            Poll::Pending
+        } else {
+            trace!("observer state: false");
+            Poll::Ready(None)
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum Error {
+    ActiveBlobNotSet,
+    InitActiveBlob(blob::Error),
+    BlobError(blob::Error),
+    WrongConfig,
+    Unitialized,
+    IO(io::Error),
+    RecordNotFound,
+    ObserverSpawnFailed(task::SpawnError),
+    WorkDirInUse,
+}
+
+impl From<blob::Error> for Error {
+    fn from(blob_error: blob::Error) -> Self {
+        Error::BlobError(blob_error)
+    }
+}
+
+/// Description
+/// Examples
+#[derive(Debug, Clone)]
+pub struct Config {
+    pub work_dir: Option<PathBuf>,
+    pub max_blob_size: Option<u64>,
+    pub max_data_in_blob: Option<u64>,
+    pub blob_file_name_prefix: Option<String>,
+    pub update_interval_ms: u64,
+}
+
+impl Default for Config {
+    fn default() -> Self {
+        Self {
+            work_dir: None,
+            max_blob_size: None,
+            max_data_in_blob: None,
+            blob_file_name_prefix: None,
+            update_interval_ms: 100,
+        }
+    }
+}
+
+async fn launch_observer<S>(spawner: S, storage: Storage)
+where
+    S: SpawnExt + Send + 'static + Unpin + Sync,
+{
+    let observer = Observer {
+        update_interval: Duration::from_millis(storage.shared.config.update_interval_ms),
+        storage: Arc::new(storage),
+        next_update: Instant::now(),
+        spawner,
+    };
+    thread::spawn(move || observer.run());
+}
+
+async fn update_active_blob(s: Arc<Storage>) -> Result<()> {
+    trace!("Storage update, lock");
+    // @TODO process unwrap explicitly
+    let mut inner = await!(s.inner.lock());
+    trace!("lock acquired");
+    let active_blob = inner.active_blob.as_ref().ok_or(Error::ActiveBlobNotSet)?;
+    let active_size = active_blob.file_size()?;
+    let config_max = s.shared.config.max_blob_size.ok_or(Error::Unitialized)?;
+    let is_full = active_size > config_max
+        || active_blob.records_count() as u64
+            >= s.shared.config.max_data_in_blob.ok_or(Error::Unitialized)?;
+
+    if is_full {
+        trace!("is full, replace");
+        let next = s.next_blob_name()?;
+        trace!("next name: {:?}", next);
+        let new_active = Blob::open_new(next)?.boxed();
```

Comment:
This operation could take a while.  Could we do this outside of lock? We can put check in one lock and if full - prepare new blob and hold lock for the second time and replace. this would be much better it terms of time we hold lock

Reply:
Done

Reply:
Is this possible at all?